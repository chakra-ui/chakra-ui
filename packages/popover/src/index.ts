export * from "./popover"
export * from "./use-popover"
