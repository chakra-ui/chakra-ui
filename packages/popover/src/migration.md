# Migration Notes

## Changes

- `returnFocusOnClose` has been changed to just `returnFocus` to keep it concise.

- `autoFocus` prop to allow users control whether the popover should automatically receive focus when it opens.
