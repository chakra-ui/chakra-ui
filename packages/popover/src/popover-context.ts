import { createContext } from "@chakra-ui/react-utils"
import { UsePopoverReturn } from "./use-popover"

export const [
  PopoverProvider,
  usePopoverContext,
] = createContext<UsePopoverReturn>({
  name: "PopoverContext",
  errorMessage:
    "usePopoverContext: `context` is undefined. Seems you forgot to wrap all popover components within `<Popover />`",
})
