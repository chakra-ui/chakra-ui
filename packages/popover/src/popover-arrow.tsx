import { chakra, HTMLChakraProps } from "@chakra-ui/system"
import { cx } from "@chakra-ui/shared-utils"
import { usePopoverContext, usePopoverStyles } from "./popover-context"

export interface PopoverArrowProps extends HTMLChakraProps<"div"> {}

export function PopoverArrow(props: PopoverArrowProps) {
  const { bg, bgColor, backgroundColor } = props
  const { getArrowProps, getArrowInnerProps } = usePopoverContext()
  const styles = usePopoverStyles()
  const arrowBg = bg ?? bgColor ?? backgroundColor
  return (
    <chakra.div
      {...getArrowProps()}
      className="chakra-popover__arrow-positioner"
    >
      <chakra.div
        className={cx("chakra-popover__arrow", props.className)}
        {...getArrowInnerProps(props)}
        __css={{
          ...styles.arrow,
          "--popper-arrow-bg": arrowBg
            ? `colors.${arrowBg}, ${arrowBg}`
            : undefined,
        }}
      />
    </chakra.div>
  )
}

PopoverArrow.displayName = "PopoverArrow"
