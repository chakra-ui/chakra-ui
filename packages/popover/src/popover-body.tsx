import { chakra, forwardRef, HTMLChakraProps } from "@chakra-ui/system"
import { cx } from "@chakra-ui/shared-utils"
import { usePopoverContext, usePopoverStyles } from "./popover-context"

export interface PopoverBodyProps extends HTMLChakraProps<"div"> {}
/**
 * PopoverBody is the main content area for the popover. Should contain
 * at least one interactive element.
 */

export const PopoverBody = forwardRef<PopoverBodyProps, "div">(
  function PopoverBody(props, ref) {
    const { getBodyProps } = usePopoverContext()

    const styles = usePopoverStyles()

    return (
      <chakra.div
        {...getBodyProps(props, ref)}
        className={cx("chakra-popover__body", props.className)}
        __css={styles.body}
      />
    )
  },
)

PopoverBody.displayName = "PopoverBody"
