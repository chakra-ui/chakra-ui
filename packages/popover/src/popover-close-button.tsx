import { CloseButton, CloseButtonProps } from "@chakra-ui/close-button"
import { forwardRef } from "@chakra-ui/system"
import { cx } from "@chakra-ui/shared-utils"
import { usePopoverContext, usePopoverStyles } from "./popover-context"

export type PopoverCloseButtonProps = CloseButtonProps

export const PopoverCloseButton = forwardRef<CloseButtonProps, "button">(
  function PopoverCloseButton(props, ref) {
    const { onClose } = usePopoverContext()
    const styles = usePopoverStyles()
    return (
      <CloseButton
        size="sm"
        onClick={onClose}
        className={cx("chakra-popover__close-btn", props.className)}
        __css={styles.closeButton}
        ref={ref}
        {...props}
      />
    )
  },
)

PopoverCloseButton.displayName = "PopoverCloseButton"
