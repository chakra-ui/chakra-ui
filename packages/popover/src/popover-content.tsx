import {
  chakra,
  forwardRef,
  HTMLChakraProps,
  SystemStyleObject,
} from "@chakra-ui/system"
import { callAll, cx, __DEV__ } from "@chakra-ui/utils"
import { usePopoverContext, usePopoverStyles } from "./popover-context"
import { PopoverTransition, PopoverTransitionProps } from "./popover-transition"

export interface PopoverContentProps extends PopoverTransitionProps {
  rootProps?: HTMLChakraProps<"div">
}

export const PopoverContent = forwardRef<PopoverContentProps, "section">(
  function PopoverContent(props, ref) {
    const { rootProps, ...contentProps } = props

    const { getPopoverProps, getPopoverPositionerProps, onAnimationComplete } =
      usePopoverContext()

    const styles = usePopoverStyles()
    const contentStyles: SystemStyleObject = {
      position: "relative",
      display: "flex",
      flexDirection: "column",
      ...styles.content,
    }

    return (
      <chakra.div
        {...getPopoverPositionerProps(rootProps)}
        __css={styles.popper}
        className="chakra-popover__popper"
      >
        <PopoverTransition
          {...getPopoverProps(contentProps, ref)}
          onAnimationComplete={callAll(
            onAnimationComplete,
            contentProps.onAnimationComplete,
          )}
          className={cx("chakra-popover__content", props.className)}
          __css={contentStyles}
        />
      </chakra.div>
    )
  },
)
if (__DEV__) {
  PopoverContent.displayName = "PopoverContent"
}
