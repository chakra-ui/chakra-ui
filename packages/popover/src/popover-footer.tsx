import { chakra, HTMLChakraProps } from "@chakra-ui/system"
import { cx } from "@chakra-ui/shared-utils"
import { usePopoverStyles } from "./popover-context"

export interface PopoverFooterProps extends HTMLChakraProps<"footer"> {}

export function PopoverFooter(props: PopoverFooterProps) {
  const styles = usePopoverStyles()
  return (
    <chakra.footer
      {...props}
      className={cx("chakra-popover__footer", props.className)}
      __css={styles.footer}
    />
  )
}

PopoverFooter.displayName = "PopoverFooter"
