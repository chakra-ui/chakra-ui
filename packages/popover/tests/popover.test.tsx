import * as React from "react"
import { fireEvent, render, wait } from "@chakra-ui/test-utils"
import { usePopover } from "../src"

const Component = () => {
  const { getTriggerProps, getPopoverProps, onClose } = usePopover()

  return (
    <div>
      <button {...getTriggerProps()}>Open</button>
      <div {...getPopoverProps()}>Popover content</div>
      <button onClick={onClose}>Close</button>
    </div>
  )
}

test("Popover renders correctly", async () => {
  const { asFragment } = render(<Component />)

  await wait()

  expect(asFragment()).toMatchSnapshot()
})

test("has proper aria attributes", async () => {
  const utils = render(<Component />)
  const trigger = utils.getByText(/open/i)

  expect(trigger).toHaveAttribute("aria-expanded", "false")
  expect(utils.getByRole("dialog", { hidden: true })).toHaveAttribute(
    "aria-hidden",
    "true",
  )

  // open the popover
  fireEvent.click(trigger)
  await wait()

  expect(trigger).toHaveAttribute("aria-haspopup", "dialog")
  expect(trigger).toHaveAttribute("aria-expanded", "true")
  expect(trigger).toHaveAttribute(
    "aria-controls",
    expect.stringContaining("popover-content"),
  )
  expect(utils.getByRole("dialog")).not.toHaveAttribute("aria-hidden")
})

test("can open and close the popover", async () => {
  const utils = render(<Component />)

  // open the popover
  fireEvent.click(utils.getByText(/open/i))
  await wait()

  // close the popover (since we can click the button, we verify that it's
  // displayed)
  fireEvent.click(utils.getByText(/close/i))
  await wait()

  // verify that content isn't displayed after closing
  expect(utils.queryByText(/content/i)).not.toBeVisible()
})

test("can close the popover by pressing escape", async () => {
  const utils = render(<Component />)

  // open the popover
  fireEvent.click(utils.getByText(/open/i))

  const dialog = await utils.findByRole("dialog")

  // close the popover with escape
  fireEvent.keyDown(dialog, { key: "Escape" })
  await wait()

  // verify popover is hidden
  // utils.getByRole("dialog", { hidden: true })
})
