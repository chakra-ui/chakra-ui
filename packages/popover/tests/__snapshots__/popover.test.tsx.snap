// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover renders correctly 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-controls="popover-content-2"
      aria-expanded="false"
      aria-haspopup="dialog"
      id="popover-trigger-2"
    >
      Open
    </button>
    <div
      aria-hidden="true"
      hidden=""
      id="popover-content-2"
      role="dialog"
      style="transform-origin: top center; position: absolute; top: 10px; left: 0px;"
      tabindex="-1"
    >
      Popover content
    </div>
    <button>
      Close
    </button>
  </div>
</DocumentFragment>
`;
